### PR TITLE
Move `PointerDataPacketConverter` from `PlatformView` to `RuntimeController`

### PR DESCRIPTION
**Pull request by dkwingsmt**
This is a refactor that moves the `PointerDataPacketConverter` from `PlatformView` to `RuntimeController`. 

This change is made for the following reasons:
- Currently, the pointer data conversion contains no platform specific logic (because the current converter's only responsibility is to make the event sequence conform Flutter's protocol). Therefore these logics should reside in a platform-independent place.
- The converter typically converts one event to many. It's better to have this conversion later than earlier.
- It removes a member from `PlatformView`, making it closer to a pure virtual class.

The reason to choose `RuntimeController` as the destination is because `RuntimeController` manages a map for views, which is required for the converter to implement a later patch https://github.com/flutter/engine/pull/51925.

## Pre-launch Checklist

- [ ] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [ ] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [ ] I read and followed the [Flutter Style Guide] and the [C++, Objective-C, Java style guides].
- [ ] I listed at least one issue that this PR fixes in the description above.
- [ ] I added new tests to check the change I am making or feature I am adding, or the PR is [test-exempt]. See [testing the engine] for instructions on writing and running engine tests.
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [ ] I signed the [CLA].
- [ ] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[C++, Objective-C, Java style guides]: https://github.com/flutter/engine/blob/main/CONTRIBUTING.md#style
[testing the engine]: https://github.com/flutter/flutter/wiki/Testing-the-engine
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

**Comment by dkwingsmt** [comment]
Update: There are some unexpected unit test failures and I'm resolving.

**Review by chunhtai — COMMENTED** [review]
is the RuntimeController to be per view? or is it going to be shared by multipleview?

**Comment by loic-sharma** [comment]
> is the RuntimeController to be per view? or is it going to be shared by multipleview?

The runtime controller is 1 per Dart VM, so it is shared by all views within a single engine instance.

File: shell/common/input_events_unittests.cc [hunk 1]
```diff
@@ -124,7 +124,9 @@ static void TestSimulatedInputEvents(
     for (int i = 0, j = 0; i < num_events; j += 1) {
       double t = j * frame_time;
       while (i < num_events && delivery_time(i) <= t) {
-        ShellTest::DispatchFakePointerData(shell.get());
+        // Use a different x every time for the pointer data converter to
+        // generate non-empty events.
```
Comment (loic-sharma) [inline shell/common/input_events_unittests.cc:128]: Should we move the "use a different x" trick to `DispatchFakePointerData`'s implementation? I don't feel strongly about this, feel free to keep as is.

File: runtime/runtime_controller.cc [hunk 2]
```diff
@@ -358,7 +358,11 @@ bool RuntimeController::DispatchPointerDataPacket(
     const PointerDataPacket& packet) {
   if (auto* platform_configuration = GetPlatformConfigurationIfAvailable()) {
     TRACE_EVENT0("flutter", "RuntimeController::DispatchPointerDataPacket");
-    platform_configuration->DispatchPointerDataPacket(packet);
+    std::unique_ptr<PointerDataPacket> converted_packet =
+        pointer_data_packet_converter_.Convert(packet);
+    if (converted_packet->GetLength() != 0) {
```
Comment (loic-sharma) [inline runtime/runtime_controller.cc:363]: I see `PlatformView::DispatchPointerDataPacket` didn't have this length check before. Could you explain why it was added?

**Comment by chunhtai** [comment]
> The runtime controller is 1 per Dart VM, so it is shared by all views within a single engine instance.

If that is the case, I think I would still prefer to put it below this level since the converter has state to keep track of pointer down or up, and each converter state should not be mixed between different view. If we put in runtime controller, we would have to maintain one converter per view.

**Comment by dkwingsmt** [comment]
The runtime controller is 1 per engine, and putting the converter under `RuntimeController` means that there's one converter per engine.

**Comment by chunhtai** [comment]
> I don't think it needs to keep states by views

If you have a gesture across multiple view, like a drag and drop, should they be treated as one gesture, or 2 gesture one for each view? sharing the same state or not will change how pointer event is synthesized

> the converter needs to be aware of views, which PlatformView doesn't provide.

Can you explain a bit more? what information from the view does converter need?

**Comment by dkwingsmt** [comment]
> Can you explain a bit more? what information from the view does converter need?

The converter simply needs to know when a view is added or removed. However, currently we managed to implement the entire multi-view rendering without adding such API to `PlatformView`. 

Well, now that I think about it, it might have been a better idea to add `Add/RemoveView` to `PlatformView` instead of to `Shell`. It feels like `Shell` is just a glue between the engine's components, not something to operate on. And in that case, `PlatformView` will have view information and this change is not needed. 
@loic-sharma What do you think?

**Comment by dkwingsmt** [comment]
> If you have a gesture across multiple view, like a drag and drop, should they be treated as one gesture, or 2 gesture one for each view? sharing the same state or not will change how pointer event is synthesized

The idea is that, a pointer's state is indexed by its pointer ID. The converter tracks whether this pointer travels across the view boundary and synthesize add/remove events accordingly. 

An easier way to think about it is what Flutter "remembers" about pointers of an app.
* Flutter should think, "I have 2 pointers, pointer 1 is currently on view 10, pointer 2 is currently not on any views."
* Flutter should not think, "I have 3 views, view 10 currently has 1 pointer on it, view 11 and view 12 don't have any pointers on it".

**Comment by loic-sharma** [comment]
> Well, now that I think about it, it might have been a better idea to add `Add/RemoveView` to `PlatformView` instead of to `Shell`. It feels like `Shell` is just a glue between the engine's components, not something to operate on. And in that case, `PlatformView` will have view information and this change is not needed.

I think the `Shell` setup we have today fits this paradigm: `Shell::AddView` just switches to the UI thread  and then calls `Engine::AddView`. It's still just glue :)

Also, the `PlatformView` is operated on the platform thread, right? Would the approach you suggested require us to track views' states in the engine on the platform thread as well? Personally I'd rather have as much pointer logic on the UI thread as possible.

**Comment by chunhtai** [comment]
> The idea is that, a pointer's state is indexed by its pointer ID. The converter tracks whether this pointer travels across the view boundary and synthesize add/remove events accordingly.

yes that can work, but then the converter will be more complex to not only memorize the pointer is down/up, it also need to know which view the pointer has been downed on.

I am not too familiar with the shell/platformview/view relationship, if it makes more sense to handle it in the shell level, then I think this is fine.

**Comment by dkwingsmt** [comment]
> then the converter will be more complex to not only memorize the pointer is down/up, it also need to know which view the pointer has been downed on.

But this is unavoidable. The native platforms typically dispatch pointer-add/remove-events when a pointer enters/leaves a view. The app would like to know it as well. So this behavior is intended. The only problem is how to organize our data to compute the logic conveniently.

**Comment by chinmaygarde** [comment]
Are we making progress on this?

**Review by chunhtai — APPROVED** [review]
Talked with @dkwingsmt offline. It makes sense to pull these kind of logic out of platform view. LGTM

**Pull request merged**